File: Code/DataStructs/BitVect.cpp
// $Id$
//
// Copyright (c) 2003-2008 greg Landrum and Rational Discovery LLC
//
//  @@ All Rights Reserved @@
//  This file is part of the RDKit.
//  The contents are covered by the terms of the BSD license
//  which is included in the file license.txt, found at the root
//  of the RDKit source tree.
//
#include "BitVect.h"
#include <sstream>
#include <limits>
#include <RDGeneral/StreamOps.h>
#include "base64.h"
#ifdef WIN32
#include <ios>
#endif
#include <cstdint>

BitVect::~BitVect(){};  // must always implement virtual destructors

void BitVect::initFromText(const char *data, const unsigned int dataLen,
                           bool isBase64, bool allowOldFormat) {
  std::stringstream ss(std::ios_base::binary | std::ios_base::in |
                       std::ios_base::out);
  if (isBase64) {
    unsigned int actualLen;
    char *decoded;
    decoded = Base64Decode((const char *)data, &actualLen);
    ss.write(decoded, actualLen);
    delete[] decoded;
  } else {
    ss.write(data, dataLen);
  }

  std::int32_t format = 0;
  std::int32_t version = 0;
  std::uint32_t nOn = 0;
  std::int32_t size;

  // earlier versions of the code did not have the version number encoded, so
  //  we'll use that to distinguish version 0
  RDKit::streamRead(ss, size);
  if (size < 0) {
    version = -1 * size;
    switch (version) {
      case 16:
        format = 1;
        break;
      case 32:
        format = 2;
        break;
      default:
        throw ValueErrorException("bad version in BitVect pickle");
        break;
    }
    RDKit::streamRead(ss, size);
  } else if (!allowOldFormat) {
    throw ValueErrorException("invalid BitVect pickle");
  }

  RDKit::streamRead(ss, nOn);
  _initForSize(static_cast<int>(size));

  // if the either have older version or or version 16 with ints for on bits
  if ((format == 0) ||
      ((format == 1) && (size >= std::numeric_limits<unsigned short>::max()))) {
    std::uint32_t tmp;
    for (unsigned int i = 0; i < nOn; i++) {
      RDKit::streamRead(ss, tmp);
      setBit(tmp);
    }
  } else if (format == 1) {  // version 16 and on bits stored as short ints
    std::uint16_t tmp;
    for (unsigned int i = 0; i < nOn; i++) {
      RDKit::streamRead(ss, tmp);
      setBit(tmp);
    }
  } else if (format == 2) {  // run length encoded format
    std::uint32_t curr = 0;
    for (unsigned int i = 0; i < nOn; i++) {
      curr += RDKit::readPackedIntFromStream(ss);
      setBit(curr);
      curr++;
    }
  }
}

File: rdkit/DataStructs/BitVect.py
# $Id$
#
#  Copyright (C) 2001-2006  greg Landrum and Rational Discovery LLC
#
#   @@ All Rights Reserved @@
#  This file is part of the RDKit.
#  The contents are covered by the terms of the BSD license
#  which is included in the file license.txt, found at the root
#  of the RDKit source tree.
#
""" This has all be re-implemented in the C++ code
"""
from __future__ import print_function
import math

class BitVect:
  def __init__(self,nBits):
    self.nBits = nBits
    self.bits = [0]*nBits

  def NumOnBits(self):
    return len(self.GetOnBits())

  def GetOnBits(self,sort=1,reverse=0):
    l = [idx for idx in xrange(self.nBits) if self.bits[idx] == 1]
    if reverse:
      l.reverse()
    return l
  
  def TanimotoSimilarity(self,other):
    if not isinstance(other,BitVect):
      raise TypeError("Tanimoto similarities can only be calculated between two BitVects")
    if len(self)!=len(other):
      raise ValueError("BitVects must be the same length")
    bc = len(self & other)
    b1 = self.NumOnBits()
    b2 = other.NumOnBits()
    return float(bc) / float(b1 + b2 - bc)
  TanimotoSimilarity = TanimotoSimilarity

  def EuclideanDistance(self,other):
    if not isinstance(other,BitVect):
      raise TypeError("Tanimoto similarities can only be calculated between two BitVects")
    bt = len(self)
    bi = len(self ^ (~ other))
    return math.sqrt(bt-bi)/bt
  
  def __getitem__(self,which):
    if which >= self.nBits or which < 0:
      raise ValueError('bad index')
    return self.bits[which]

  def __setitem__(self,which,val):
    if which >= self.nBits or which < 0:
      raise ValueError('bad index')
    if val not in [0,1]:
      raise ValueError('val must be 0 or 1')

    self.bits[which] = val
      
  def __len__(self):
    return self.nBits

  def __and__(self,other):
    if not isinstance(other,BitVect):
      raise TypeError("BitVects can only be &'ed with other BitVects")
    if len(self) != len(other):
      raise ValueError("BitVects must be of the same length")

    l1 = self.GetOnBits()
    l2 = other.GetOnBits()
    r = [bit for bit in l1 if bit in l2]
    return r

  def __or__(self,other):
    if not isinstance(other,BitVect):
      raise TypeError("BitVects can only be |'ed with other BitVects")
    if len(self) != len(other):
      raise ValueError("BitVects must be of the same length")
    l1 = self.GetOnBits()
    l2 = other.GetOnBits()
    r = l1 + [bit for bit in l2 if bit not in l1]
    r.sort()
    return r

  def __xor__(self,other):
    if not isinstance(other,BitVect):
      raise TypeError("BitVects can only be ^'ed with other BitVects")
    if len(self) != len(other):
      raise ValueError("BitVects must be of the same length")
    
    l1 = self.GetOnBits()
    l2 = other.GetOnBits()
    r = [bit for bit in l1 if bit not in l2] + [bit for bit in l2 if bit not in l1]
    r.sort()
    return r

  def __invert__(self):
    res = BitVect(len(self))
    for i in xrange(len(self)):
      res[i] = not self[i]
    return res
  
class SparseBitVect(BitVect):
  def __init__(self,nBits):
    self.nBits = nBits
    self.bits = []

  def NumOnBits(self):
    return len(self.bits)

  def GetOnBits(self,sort=1,reverse=0):
    l = self.bits[:]
    if sort:
      l.sort()
    if reverse:
      l.reverse()
    return l
  
  def __getitem__(self,which):
    if which >= self.nBits or which < 0:
      raise ValueError('bad index')
    if which in self.bits:
      return 1
    else:
      return 0

  def __setitem__(self,which,val):
    if which >= self.nBits or which < 0:
      raise ValueError('bad index')
    if val == 0:
      if which in self.bits:
        self.bits.remove(which)
    else:
      self.bits.append(which)
      
  def __len__(self):
    return self.nBits

if __name__ == '__main__':
  b1 = BitVect(10)
  b2 = SparseBitVect(10)
  b1[0] = 1
  b2[0] = 1
  b1[3] = 1
  b2[4] = 1
  b2[5] = 1
  b2[5] = 0
  print('b1:',b1.GetOnBits())
  print('b2:',b2.GetOnBits())
  print('&:', b1 & b2)
  print('|:', b1 | b2)
  print('^:', b1 ^ b2)
  print('b1.Tanimoto(b2):',b1.TanimotoSimilarity(b2))
  print('b1.Tanimoto(b1):',b1.TanimotoSimilarity(b1))
  print('b2.Tanimoto(b2):',b2.TanimotoSimilarity(b2))
  print('b2.Tanimoto(b1):',b2.TanimotoSimilarity(b1))
  
  
